### ISSUE TITLE
SourceRoot should be able to delete files.

### ISSUE THREAD

**Issue by matozoid** [javaparser/javaparser#1038]
... in case of a dead code remover, for example.

**Comment by ryan-beckett** [comment]
Do CompilationUnits have the ability to rewrite their respective files?

**Comment by matozoid** [comment]
Yes, check out the `storage` field. It won't help much for this issue, since `SourceRoot` will simply rewrite the file when you call `saveAll()`. So I was looking for some way to flag a CU for deletion (or skip flagging and delete right away, and delete from the cache too.)

Okay, that's for the non-callback functionality. For the callback we have a return enum containing SAVE and DONT_SAVE. It would be easy to add DELETE there.

So I guess this is two issues in one :-)

**Comment by ryan-beckett** [comment]
@matozoid: Maybe you can shed some light on this. In the code below, the `CompilationUnit` produced as a result of parsing the file at `absolutePath`, has its storage set with `absolutePath`. However, if the result of the callback is `SAVE`, the unit is saved with a storage that points to `path`. Is this a mistake? 


```
final Path path = packageAbsolutePath(root, startPackage);
        Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
            @Override
            public FileVisitResult visitFile(Path absolutePath, BasicFileAttributes attrs) throws IOException {
                if (!attrs.isDirectory() && absolutePath.toString().endsWith(".java")) {
                    Path localPath = root.relativize(absolutePath);
                    Log.trace("Parsing %s", localPath);
                    final ParseResult<CompilationUnit> result = javaParser.parse(COMPILATION_UNIT,
                            provider(absolutePath));
                    result.getResult().ifPresent(cu -> cu.setStorage(absolutePath));
                    if (callback.process(localPath, absolutePath, result) == SAVE) {
                        if (result.getResult().isPresent()) {
                            save(result.getResult().get(), path);
                        }
                    }
                }
                return FileVisitResult.CONTINUE;
            }
        });
```


Why not simply say:

```
if (callback.process(localPath, absolutePath, result) == SAVE) {
        result.getResult().ifPresent(cu -> cu.getStorage().get().save(printer);
        }
}
```

**Comment by ryan-beckett** [comment]
> So I was looking for some way to flag a CU for deletion (or skip flagging and delete right away, and delete from the cache too.)

@matozoid: I've spent a few hours thinking about this, and I can't think of an elegant solution. 

I was able to add the ability to delete files when using the callback methods. If you're interested in adding that only, I'll submit a PR.

**Comment by matozoid** [comment]
@ryan-beckett well, that's at least half the problem solved :-) Do submit a PR.

**Comment by matozoid** [comment]
@ryan-beckett your remark about saving seems to be correct. `Storage` was added after `SourceRoot` so you may see more code around that seems to think `Storage` does not exist.

**Comment by matozoid** [comment]
@ryan-beckett about an elegant solution for the cached version: I think you start with adding a `delete` method to `SourceRoot` that takes a `CompilationUnit`. Then you can choose to do the right-away solution (which is less desirable because it suddenly changes the read-only view of the source directory - suddenly you can make files disappear in the middle of processing!) or find some kind of flag (check out the data field on Node, you could add a `ScheduledForDeletion` object to it or so) and look for it in the `save` methods. When it is set, do not save but delete.

**Comment by ryan-beckett** [comment]
@matozoid: I'll submit it after work. I was thinking the exact same thing regarding the `delete` method, but I feel like it's such an arkward design choice, not to even consider it. Maybe it's time to consult with StackOverflow!

**Comment by matozoid** [comment]
If you like, not sure what random people can add here though :-)
